[dynamo] support builtins on dynamic shapes, e.g. builtins.neg

bdhirsh opened the pull request:
Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92643
* __->__ #92618
* #92617



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

bdhirsh committed:
[dynamo] support builtins on dynamic shapes, e.g. builtins.neg

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -233,7 +240,11 @@ def call_function(
+            # For those that don't, we should be able to rely on SymInt erroring.
+            # (Alternative would be to directly read off the list of allowed SymInt operations,
+            # although I'm not sure what that buys us).
+            and (tensor_args or dyn_shape_args)
```
My thought was that for builtins that we allow in the FX graph (e.g. builtins.neg), we'd ideally want to be able to reuse the same dynamo code for tensors and dynamic shapes as much as possible.

The only real difference (as I see it) is that SymInt supports a smaller list of builtins than Tensor - but I'm just relying on SymInt erroring properly. E.g. when I tried `x.shape[0] or x.shape[1]` with this PR, I get:
```
TypeError: unsupported operand type(s) for |: 'SymInt' and 'SymInt'
```

ezyang replied:
The particular translation you quote here looks wrong. SymInt could in principle support bitwise or, and this would be completely the wrong thing to call if the user wrote logical or.

ezyang commented:
Unfortunately, I don't have a good enough grasp on this part of dynamo to know automatically what the right structure is. Maybe @voznesenskym or @jansel have an opinion

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -278,6 +289,13 @@ def call_function(
                         need_unwrap=need_unwrap,
                         **options,
                     )
+                elif all(isinstance(x, DynamicShapeVariable) for x in args):
```
This doesn't seem very robust as if one of the args is a ConstantVariable it will fall through to the next case.

Perhaps we could inspect the type of `proxy` instead.  I am assuming in this case it won't actually be an FX proxy.

bdhirsh committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
I think it would make the most sense to land @jbschlosser's https://github.com/pytorch/pytorch/pull/93271 first, and then build this on top of that (neg is not reversible, but there's still some common structure)

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support builtins on dynamic shapes, e.g. builtins.neg"

Support code like this:
```
def f(x):
    y = x.shape[0]
    return -y
```

That calls the builtin `builtins.neg(DynamicShape)` in dynamo.

We have support for builtins on tensors but not dynamic shapes. It seems like there are a couple of ways to get this to work, so very open to feedback / if someone thinks there's a better way.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
